pkg/gather/gather.go

```diff
@@ -0,0 +1,256 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gather
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	"github.com/submariner-io/submariner-operator/internal/cli"
+	"github.com/submariner-io/submariner-operator/internal/component"
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/brokercr"
+	"github.com/submariner-io/submariner-operator/pkg/client"
+	"github.com/submariner-io/submariner-operator/pkg/cluster"
+	"github.com/submariner-io/submariner-operator/pkg/names"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Options struct {
+	Type                 string
+	Module               string
+	Directory            string
+	IncludeSensitiveData bool
+}
+
+const (
+	Logs      = "logs"
+	Resources = "resources"
+)
+
+var ModuleFlags = map[string]bool{
+	component.Connectivity:     false,
+	component.ServiceDiscovery: false,
+	component.Broker:           false,
+	component.Operator:         false,
+}
+
+var TypeFlags = map[string]bool{
+	"logs":      false,
+	"resources": false,
+}
+
+var gatherFuncs = map[string]func(string, Info) bool{
+	component.Connectivity:     gatherConnectivity,
+	component.ServiceDiscovery: gatherDiscovery,
+	component.Broker:           gatherBroker,
+	component.Operator:         gatherOperator,
+}
+
+func Data(clusterInfo *cluster.Info, status reporter.Interface, options Options, writer io.Writer) bool {
+	if options.Directory == "" {
+		options.Directory = "submariner-" + time.Now().UTC().Format("20060102150405") // submariner-YYYYMMDDHHMMSS
+	}
+
+	if _, err := os.Stat(options.Directory); os.IsNotExist(err) {
+		err := os.MkdirAll(options.Directory, 0o700)
+		if err != nil {
+			status.Failure("Error creating directory %q: %s", options.Directory, err)
+			return false
+		}
+	}
+
+	gatherDataByCluster(clusterInfo, status, options, writer)
+
+	_, _ = fmt.Fprintf(writer, "Files are stored under directory %q\n", options.Directory)
+
+	return true
+}
+
+func gatherDataByCluster(clusterInfo *cluster.Info, status reporter.Interface, options Options, writer io.Writer) {
+	var err error
+	clusterName := clusterInfo.Name
+
+	_, _ = fmt.Fprintf(writer, "Gathering information from cluster %q\n", clusterName)
```

Comment:
Why are you using an `io.Writer` for progress output and not the `reporter.Interface`?

Reply:
I did use `reporter.Interface` but was advised to not change it [here](https://github.com/submariner-io/submariner-operator/pull/1801#discussion_r794659155). The idea is to give options to users to log these informational info anywhere they want.

Reply:
Providing two avenues for users to log info would be confusing. This works for subctl b/c it all goes to `fmt.Printf`. This is another reason why I think `gather` is specific to subctl.

Reply:
I think there are plans to use `gather` in ACM UI as well, isn't it @vthapar ?

Reply:
@Jaanki We haven't yet figured out how to make gather work from UI. Note that UI doesn't use Go and can't use Go APIs. Similar to diagnose we will have to run a pod in cluster to run gather and then make results [maybe a tar file] available to UI. Most likely it would be running `subctl gather`, not much different from what we're planning with diagnose.

I kinda agree with @tpantelis There might be some use case to provide API to gather IP tables etc. but for pod logs and resources not much. We can park this requirement unless we come up with actual use case, whether it is from UI or otherwise. Don't want to do some changes which still don't work for UI and end up being dead code.